src/lambertw.jl

```diff
@@ -0,0 +1,358 @@
+
+Compute the `k`th branch of the Lambert W function of `z`. If `z` is real, `k` must be
+either `0` or `-1`. For `Real` `z`, the domain of the branch `k = -1` is `[-1/e,0]` and the
+domain of the branch `k = 0` is `[-1/e,Inf]`. For `Complex` `z`, and all `k`, the domain is
+the complex plane. When using root finding to compute `W`, a value for `W` is returned
+with a warning if it has not converged after `maxits` iterations.
+
+```jldoctest
+julia> lambertw(-1/e,-1)
+-1.0
+
+julia> lambertw(-1/e,0)
+-1.0
+
+julia> lambertw(0,0)
+0.0
+
+julia> lambertw(0,-1)
+-Inf
+
+julia> lambertw(Complex(-10.0,3.0), 4)
+-0.9274337508660128 + 26.37693445371142im
+```
+
+"""
+lambertw(z, k::Integer=0, maxits::Integer=1000) = lambertw_(z, k, maxits)
+
+function lambertw_(x::Real, k, maxits)
+    k == 0 && return lambertwk0(x, maxits)
+    k == -1 && return lambertwkm1(x, maxits)
+    throw(DomainError(k, "lambertw: real x must have branch k == 0 or k == -1"))
+end
+
+function lambertw_(x::Union{Integer,Rational}, k, maxits)
+    if k == 0
+        x == 0 && return float(zero(x))
+        x == 1 && return convert(typeof(float(x)), omega) # must be a more efficient way
+    end
+    return lambertw_(float(x), k, maxits)
+end
+
+### Complex z ###
+
+# choose initial value inside correct branch for root finding
+function lambertw_(z::Complex{T}, k, maxits) where T<:Real
+    one_t = one(T)
+    local w::Complex{T}
+    pointseven = 7//10
+    if abs(z) <= one_t/convert(T,MathConstants.e)
+        if z == 0
+            k == 0 && return z
+            return complex(-convert(T,Inf),zero(T))
+        end
+        if k == 0
+            w = z
+        elseif k == -1 && imag(z) == 0 && real(z) < 0
+            w = complex(log(-real(z)),1//10^7) # need offset for z ≈ -1/e.
+        else
+            w = log(z)
+            k != 0 ? w += complex(0,k * 2 * pi) : nothing
+        end
+    elseif k == 0 && imag(z) <= pointseven && abs(z) <= pointseven
+        w = abs(z+ 1//2) < 1//10 ? imag(z) > 0 ? complex(pointseven,pointseven) : complex(pointseven,-pointseven) : z
+    else
+        if real(z) == convert(T,Inf)
+            k == 0 && return z
+            return z + complex(0,2*k*pi)
+        end
+        real(z) == -convert(T,Inf) && return -z + complex(0,(2*k+1)*pi)
+        w = log(z)
+        k != 0 ? w += complex(0, 2*k*pi) : nothing
+    end
+    return _lambertw(z, w, maxits)
+end
+
+lambertw_(z::Complex{T}, k, maxits) where T<:Integer = lambertw_(float(z), k, maxits)
+lambertw_(n::Irrational, k, maxits) = lambertw_(float(n), k, maxits)
+
+# lambertw(e + 0im,k) is ok for all k
+# Maybe this should return a float. But, this should cause no type instability in any case
+function lambertw_(::typeof(MathConstants.e), k, maxits)
+    k == 0 && return 1
+    throw(DomainError(k))
+end
+
+### omega constant ###
+
+const omega_const_ = 0.567143290409783872999968662210355
+# The BigFloat `omega_const_bf_` is set via a literal in the function __init__ to prevent a segfault
```

Comment:
I wonder if the issue (segfault) that required `omega_const_bf` initialization in `__init__()` is resolved, and it could be declared here:
```suggestion
### Omega constant: Ω exp(Ω) = 1 (Ω = lambertw(1)) ###

const Omega_constant = 0.567143290409783872999968662210355
const Omega_constant_BigFloat = BigFloat("0.5671432904097838729999686622103555497538157871865125081351310792230457930866845666932194")
```

Or maybe Omega constant could be moved to `IrrationalConstants.jl`?

Reply:
The "workaround" for the segfault was the standard way to deal with this problem. AFAIK, it has not changed. This is work that has to be done after the dynamic library (GMP, or something like that in this case) is linked.

Reply:
I see. The potential alternative that does not implicate `__init__()` could be
```julia
const Omega_BigFloat256 = Ref{BigFloat}()

@irrational Omega 0.567143290409783872999968662210355 Omega_constant()

# compute BigFloat Omega constant at arbitrary precision
function Omega_constant()
    isassigned(Omega_BigFloat256) || (Omega_BigFloat256[] = BigFloat("0.5671432904097838729999686622103555497538157871865125081351310792230457930866845666932194"))
    oc = Omega_BigFloat256[]
    precision(BigFloat) <= 256 && return oc
    myeps = eps(BigFloat)
    for i in 1:100
        nextoc = (1 + oc) / (1 + exp(oc))
        abs(oc - nextoc) <= myeps && break
        oc = nextoc
    end
    return oc
end
```